src/kbmod/search/raw_image_eigen.cpp

```diff
@@ -0,0 +1,767 @@
+    Image block = image.block(idx.i, idx.j, dim, dim);
+    RawImageEigen stamp(block, obstime);
+
+    if (interpolate) {
+      for (int yoff = 0; yoff < dim; ++yoff) {
+        for (int xoff = 0; xoff < dim; ++xoff) {
+          // double check if it was this->interpolate or stamp.interpolate
+          stamp.image(yoff, xoff) = this->interpolate(x + static_cast<float>(xoff - radius),
+                                                      y + static_cast<float>(yoff - radius));
+        }
+      }
+    }
+
+    return stamp;
+  }
+
+
+  void RawImageEigen::add(const float x, const float y, const float value) {
+    // In the original implementation this just does array[x, y] += value
+    // even though that doesn't really make any sense? The cast is probably
+    // implicit. This is likely slower than before, because the floor is used
+    // not a cast. Floor is more correct, but I added the constructor nonetheless
+    Index p(x, y, false);
+    if (p.is_within(width, height))
+      image(p.i, p.j) += value;
+  }
+
+
+  void RawImageEigen::add(const unsigned x, const unsigned y, const float value) {
+    Index p(x, y);
+    if (p.is_within(width, height))
+      image(p.i, p.j) += value;
+  }
+
+
+  void RawImageEigen::add(const Index p, const float value) {
+    if (p.is_within(width, height))
+      image(p.i, p.j) += value;
+  }
+
+
+  void RawImageEigen::interpolated_add(const float x, const float y, const float value) {
+    // interpolation weights and neighbors
+    auto iwn = get_interp_neighbors_and_weights(x, y);
+    add(iwn.top_left,  value * iwn.w_tl);
+    add(iwn.top_right, value * iwn.w_tr);
+    add(iwn.bot_left,  value * iwn.w_bl);
+    add(iwn.bot_right, value * iwn.w_br);
+  }
+
+
+  std::array<float, 2> RawImageEigen::compute_bounds() const {
+    float min_val = FLT_MAX;
+    float max_val = -FLT_MAX;
+
+    for (auto elem : image.reshaped())
+      if (elem != NO_DATA) {
+        min_val = std::min(min_val, elem);
+        max_val = std::max(max_val, elem);
+      }
+
+    // Assert that we have seen at least some valid data.
+    assert(max_val != -FLT_MAX);
+    assert(min_val != FLT_MAX);
+
+    return {min_val, max_val};
+  }
+
+
+  // multiple times now I've encountered the issue where I can't reduce
+  // simple array operations to method calls in Eigen because masking is
+  // baked into the image in a way that it doesn't represent the identity
+  // of the operation in question. This forces us to unravel the arrays
+  // in for loops and check if NO_DATA. I wonder if resorting to Eigens
+  // expressions would be faster than this. I also wonder how "dangerous"
+  // it would be. How often, let's say, do we expect to see exactly 0 in real data?
+  // Then convolution could be as easy as:
+  // Image zeroed = (array.array() == NO_DATA).select(0, array.array());
+  // result(i, j) = zeroed.block<krows, kcols>(i, j).cwiseProduct(kernel).sum()
```

Comment:
I'm not sure I understand what you meant. Let's say:

$$\mathrm{PSF} = \left[\begin{matrix}
1 & 2 & 3 \\
4 & 5 & 6 \\
7 & 8 & 9
\end{matrix}\right]$$

and a block of image:

$$\mathrm{IMG} = \left[\begin{matrix}
0 & 2 & 3 \\
4 & 0 & 6 \\
7 & 8 & 0
\end{matrix}\right]$$

then 

$$ \left[ \mathrm{IMG}  \circledast  \mathrm{PSF} \right]_{1, 1}  = (0\cdot 1) + 2\cdot 2 + 3\cdot 3 + 4\cdot 4 + (5\cdot 0) + 6\cdot 6 + 7\cdot 7 + 8\cdot 8 + (9\cdot 0) $$

so how is having `NO_DATA` set to 0 different? Let's say we leave out `NO_DATA`:

$$\mathrm{IMG} = \left[\begin{matrix}
\mathrm{ND} & 2 & 3 \\
4 & \mathrm{ND} & 6 \\
7 & 8 & \mathrm{ND}
\end{matrix}\right]$$


$$ \left[ \mathrm{IMG}  \circledast  \mathrm{PSF} \right]_{1, 1}  =( ) + 2\cdot 2 + 3\cdot 3 + 4\cdot 4 + ( ) + 6\cdot 6 + 7\cdot 7 + 8\cdot 8 +  ( )$$

The psf sum then can be something like `(block == NO_DATA).(0, psf).sum()` but I don't think it's critical to renormalize with just the PSF. I don't think much would change even if we didn't renormalize. By definition convolution omits the normalization factor, but really, perhaps even if we just renormalized with total sum of PSF it wouldn't be the worst because it down-weighs results near things we don't want in the first place.

Reply:
The difference is in the normalization. If it is not important to normalize the PSF, then we can use the zero approach.